consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,488 @@
+receive broadcast messages and there is no mechanism for these messages to be
+re-sent.
+
+We refer to a consensus instance to mean consensus for a specific height,
+consensus instances are independent (they have no shared state), that is
+messages or values from the previous height have no bearing on the outcome of
+the current height
+
+We assume the existence of an application that feeds values to be agreed upon
+to the consensus instance and also receives agreed upon values from the
+consensus. The application also provides application specific implementations
+for certain functions that the consensus algorithm relies upon, such as a
+function that determines the validity of a value. 
+
+### Algorithm
+
+See supporting [functions](#Functions) and [notation](#Appendix-1-Notation).
+
+Note all state modifications are contained in the following code block,
+supporting functions cannot modify state, they all operate without side
+effects.
+
+Variable names                  |Instance state                          
+--------------------------------|----------------------------------------
+`H - height`                    |`Hc - current height`                                      
+`R - round          `           |`Rc - current round`                    
+`V - value`                     |`Rd - desired round`                    
+`T - Message type`              |`Vc - currently proposed value`         
+`RCC - round change certificate`|`PCc - current prepared certificate`    
+`PC - prepared certificate`     |`Sc - current participant state`        
+`S - participant state`         |
+`M or N - message sets`         |
+
+```
+
+// Upon receiving this event the algorithm transitions to the next height the
+// final committed event signifies that the application has accepted the agreed
+// upon value for the current height.
+upon: <FinalCommittedEvent>
+  Hc ← Hc+1
+  Rc ← 0
+  Rd ← 0
+  Sc ← AcceptRequest
+  Vc ← nil
+  schedule onRoundChangeTimeout(Hc, 0) after roundChangeTimeout(0)
+
+// A request event is a request to reach agreement on the provided value, the
+// request event is sent by the application, if this parcicipant is the proposer
+// it will propose that value by sending a preprepared message.
+upon: <RequestEvent, Hc, V> && Sc = AcceptRequest
+  if Rc = 0 && isProposer(Hc, Rd) {
+    bc(<Preprepare, Hc, 0, V, nil>)
+  }
+
+// When receiving a preprepare from a proposer participants will vote for the
+// value (if valid) by sending a prepare message.
+upon: <Preprepare, Hc, Rd, V, RCC> from proposer(Hc, Rd) && Sc = AcceptRequest
+  if (Rd > 0 && validRCC(V, RCC)) || (Rd = 0 && RCC = nil)  {
+    Rc ← Rd
+    Vc ← V
+    Sc ← Preprepared
+    bc(<Prepare, Hc, Rd, Vc>)
+  }
+
+// When a participant sees at least 2f+1 prepare or commit messages for a value
+// it will send a commit message for that value.
+upon: M ← { <T, Hc, Rd, Vc> : T ∈ {Prepare, Commit} } && |M| >= 2f+1 && Sc ∈ {AcceptRequest, Preprepared} 
```

Comment:
shouldn't we be already in `Preprepared`?

Reply:
Yes I think so, thanks!